File: fop-0_20_2-maintain/src/org/apache/fop/layout/hyphenation/ByteVector.java
/*
 * $Id$
 * ============================================================================
 *                    The Apache Software License, Version 1.1
 * ============================================================================
 * 
 * Copyright (C) 1999-2003 The Apache Software Foundation. All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without modifica-
 * tion, are permitted provided that the following conditions are met:
 * 
 * 1. Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 * 
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 * 
 * 3. The end-user documentation included with the redistribution, if any, must
 *    include the following acknowledgment: "This product includes software
 *    developed by the Apache Software Foundation (http://www.apache.org/)."
 *    Alternately, this acknowledgment may appear in the software itself, if
 *    and wherever such third-party acknowledgments normally appear.
 * 
 * 4. The names "FOP" and "Apache Software Foundation" must not be used to
 *    endorse or promote products derived from this software without prior
 *    written permission. For written permission, please contact
 *    apache@apache.org.
 * 
 * 5. Products derived from this software may not be called "Apache", nor may
 *    "Apache" appear in their name, without prior written permission of the
 *    Apache Software Foundation.
 * 
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESSED OR IMPLIED WARRANTIES,
 * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
 * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * APACHE SOFTWARE FOUNDATION OR ITS CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLU-
 * DING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
 * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 * ============================================================================
 * 
 * This software consists of voluntary contributions made by many individuals
 * on behalf of the Apache Software Foundation and was originally created by
 * James Tauber <jtauber@jtauber.com>. For more information on the Apache
 * Software Foundation, please see <http://www.apache.org/>.
 */ 
package org.apache.fop.layout.hyphenation;

import java.io.Serializable;

/**
 * This class implements a simple byte vector with access to the
 * underlying array.
 *
 * @author Carlos Villegas <cav@uniscope.co.jp>
 */
public class ByteVector implements Serializable {

    /**
     * Capacity increment size
     */
    private static final int DEFAULT_BLOCK_SIZE = 2048;
    private int BLOCK_SIZE;

    /**
     * The encapsulated array
     */
    private byte[] array;

    /**
     * Points to next free item
     */
    private int n;

    public ByteVector() {
        this(DEFAULT_BLOCK_SIZE);
    }

    public ByteVector(int capacity) {
        if (capacity > 0)
            BLOCK_SIZE = capacity;
        else
            BLOCK_SIZE = DEFAULT_BLOCK_SIZE;
        array = new byte[BLOCK_SIZE];
        n = 0;
    }

    public ByteVector(byte[] a) {
        BLOCK_SIZE = DEFAULT_BLOCK_SIZE;
        array = a;
        n = 0;
    }

    public ByteVector(byte[] a, int capacity) {
        if (capacity > 0)
            BLOCK_SIZE = capacity;
        else
            BLOCK_SIZE = DEFAULT_BLOCK_SIZE;
        array = a;
        n = 0;
    }

    public byte[] getArray() {
        return array;
    }

    /**
     * return number of items in array
     */
    public int length() {
        return n;
    }

    /**
     * returns current capacity of array
     */
    public int capacity() {
        return array.length;
    }

    public void put(int index, byte val) {
        array[index] = val;
    }

    public byte get(int index) {
        return array[index];
    }

    /**
     * This is to implement memory allocation in the array. Like malloc().
     */
    public int alloc(int size) {
        int index = n;
        int len = array.length;
        if (n + size >= len) {
            byte[] aux = new byte[len + BLOCK_SIZE];
            System.arraycopy(array, 0, aux, 0, len);
            array = aux;
        }
        n += size;
        return index;
    }

    public void trimToSize() {
        if (n < array.length) {
            byte[] aux = new byte[n];
            System.arraycopy(array, 0, aux, 0, n);
            array = aux;
        }
    }

}

File: dotnet/src/Fop.Hyphenation/ByteVector.cs
// Licensed to the Apache Software Foundation (ASF) under one or more
// contributor license agreements.  See the NOTICE file distributed with
// this work for additional information regarding copyright ownership.
// The ASF licenses this file to You under the Apache License, Version 2.0
// (the "License"); you may not use this file except in compliance with
// the License.  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

namespace Fop.Hyphenation;

/// <summary>
/// A simple growable byte vector with access to the underlying array.
/// <para>
/// Faithful port of <c>org.apache.fop.hyphenation.ByteVector</c>, preserving its
/// explicit capacity (block-size) growth strategy.
/// </para>
/// </summary>
public class ByteVector
{
    /// <summary>Capacity increment size.</summary>
    private const int DefaultBlockSize = 2048;

    private readonly int blockSize;

    /// <summary>The encapsulated array.</summary>
    private byte[] array;

    /// <summary>Points to next free item.</summary>
    private int n;

    /// <summary>Construct a byte vector instance with the default block size.</summary>
    public ByteVector()
        : this(DefaultBlockSize)
    {
    }

    /// <summary>Construct a byte vector instance.</summary>
    /// <param name="capacity">initial block size.</param>
    public ByteVector(int capacity)
    {
        blockSize = capacity > 0 ? capacity : DefaultBlockSize;
        array = new byte[blockSize];
        n = 0;
    }

    /// <summary>
    /// Construct a byte vector instance over an existing array.
    /// </summary>
    /// <param name="a">byte array to use.</param>
    /// <remarks>
    /// Faithful quirk: matching the Java original, the length pointer is
    /// initialized to 0 here (not <c>a.Length</c>), unlike <see cref="CharVector"/>.
    /// </remarks>
    public ByteVector(byte[] a)
    {
        ArgumentNullException.ThrowIfNull(a);
        blockSize = DefaultBlockSize;
        array = a;
        n = 0;
    }

    /// <summary>Construct a byte vector instance over an existing array.</summary>
    /// <param name="a">byte array to use.</param>
    /// <param name="capacity">initial block size.</param>
    public ByteVector(byte[] a, int capacity)
    {
        ArgumentNullException.ThrowIfNull(a);
        blockSize = capacity > 0 ? capacity : DefaultBlockSize;
        array = a;
        n = 0;
    }

    /// <summary>Obtain the byte vector array.</summary>
    public byte[] GetArray() => array;

    /// <summary>Obtain the number of items in the array.</summary>
    public int Length => n;

    /// <summary>Obtain the capacity of the array.</summary>
    public int Capacity => array.Length;

    /// <summary>Put a byte at the given index.</summary>
    /// <param name="index">the index.</param>
    /// <param name="val">a byte.</param>
    public void Put(int index, byte val) => array[index] = val;

    /// <summary>Get the byte at the given index.</summary>
    /// <param name="index">the index.</param>
    public byte Get(int index) => array[index];

    /// <summary>
    /// Allocate memory in the array, like <c>malloc()</c>.
    /// </summary>
    /// <param name="size">size to allocate.</param>
    /// <returns>the previous length (offset of the allocation).</returns>
    public int Alloc(int size)
    {
        int index = n;
        int len = array.Length;
        if (n + size >= len)
        {
            byte[] aux = new byte[len + blockSize];
            Array.Copy(array, 0, aux, 0, len);
            array = aux;
        }

        n += size;
        return index;
    }

    /// <summary>Trim the byte vector to its current length.</summary>
    public void TrimToSize()
    {
        if (n < array.Length)
        {
            byte[] aux = new byte[n];
            Array.Copy(array, 0, aux, 0, n);
            array = aux;
        }
    }
}
